### PR TITLE
Fix LigerCrossEntropyLoss Reduction Behavior for "None" Mode

### PR DESCRIPTION
**Pull request by hebiao064**
## Summary

Closes https://github.com/linkedin/Liger-Kernel/issues/421 

This pull request addresses an issue in the `cross_entropy_forward` function where the `reduction="none"` mode did not behave as expected. 

Previously, the function always returned a single scalar value, even when reduction="none" was specified. This update ensures that when reduction="none" is used, the function directly outputs the unreduced loss array (loss_1d) instead of summing it.

### Changes Made:

- Added a condition to handle `reduction="none"`, ensuring the function outputs loss_1d directly.
- Updated the computation of z_loss to respect the reduction="none" mode.
- Add test for cases when `reduction="none"`

#### Background on Gradients in PyTorch

- **Scalar Outputs**: When a tensor is a scalar (a single number), PyTorch can compute gradients automatically by assuming the scalar has an implicit gradient of 1.0.
- **Non-Scalar Outputs**: For tensors that are not scalars, gradients must be provided explicitly because PyTorch cannot infer the shape or distribution of gradients. Without this, it raises the error: "grad can be implicitly created only for scalar outputs."

#### Why reduction="none" Needs Explicit Gradients

When `reduction="none"`, the loss function does not reduce the per-example loss values into a single scalar. Instead, it outputs a vector of losses, with one value per example in the batch. This means that the loss tensor has multiple values, and PyTorch cannot assume what the gradient for each of these values should be unless explicitly provided.

#### The Fix
By passing `gradient=torch.ones_like(loss)` to `backward()`:

- **Gradient Tensor**: The `torch.ones_like(loss)` serves as the gradient tensor. It specifies that each element in the loss tensor contributes equally to the gradients during backpropagation.
- **Shape Match**: The gradient tensor's shape matches the loss tensor's shape, fulfilling PyTorch's requirements for non-scalar outputs during backward().

## Testing Done

make test

`pytest /home/jobuser/Liger-Kernel/test/transformers/test_cross_entropy.py` shows:

```
=================================== 93 passed, 1 warning in 13.18s ===================================
```

- Hardware Type: NVIDIA A100-SXM4-80GB 
- [x] run `make test` to ensure correctness
- [x] run `make checkstyle` to ensure code style
- [x] run `make test-convergence` to ensure convergence

### CODE REVIEW

**Comment by Tcc0403** [comment]
LGTM!

File: src/liger_kernel/ops/cross_entropy.py [hunk 1]
```diff
@@ -285,7 +285,11 @@ def cross_entropy_forward(
...
     )
 
-    loss = torch.sum(loss_1d)
+    if reduction == "none":
+        loss = loss_1d
+    else:
+        loss = torch.sum(loss_1d)
+
     if return_z_loss == _TRUE.value:
         z_loss = torch.sum(z_loss_1d)
```
Comment (Tcc0403) [inline src/liger_kernel/ops/cross_entropy.py:294]: Let's add none reduction option for z_loss as well

Comment (hebiao064) [reply src/liger_kernel/ops/cross_entropy.py]: Thanks for pointing it out, just updated, pls review.

**Review by Tcc0403 — APPROVED** [review]

**Comment by ByronHsu** [comment]
Awesome work! Please push branch to the main repo next time, so you can run CI directly. We have disabled CI from the outside fork
